Port `prod` kernel to structured kernels.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62711 Port `norm` kernel to structured kernels.
* **#62024 Port `prod` kernel to structured kernels.**
* #61643 Port `mean` kernel to structured kernels.

Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

ysiraichi committed:
Port `prod` kernel to structured kernels.

Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29847122).

ezyang commented:
rebasing this stack; @ysiraichi you'll need to ghstack checkout after this

ezyang committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29847122).

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29847122).

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ysiraichi commented:
Rebased.

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29847122).

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1027,21 +1027,35 @@ Tensor trace_cpu(const Tensor& self) {
+  auto shape = meta::get_reduction_shape(self, {}, false);
+  Tensor result = at::empty(shape, self.options().dtype(dtype));
+  impl_func_prod(self, {}, false, dtype, result);
+  return result;
```
This looks suspicious. How come we have to define `prod` here?

ysiraichi replied:
The problem is that we can't call `prod.dim_int` since it expects only one dimension to reduce (`prod`reduces all of them). This differs from `sum.dim_IntList`, that expects many dimensions: `int[1]` (i.e. we can call `sum.dim_IntList` inside `sum` kernel)

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -81,76 +116,55 @@ ScalarType check_allany_and_get_output_dtype(
-    bool keepdim) {
-  DimVector shape;
-
+    const c10::optional<int64_t>& dim) {
```
cool, I definitely like the new signature better

ezyang committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29847122).

ysiraichi committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29847122).

ezyang committed:
Update on "Port `prod` kernel to structured kernels."


Tracking issue: #55070

In this PR, I also broke down the meta functions of other reduction kernels (e.g. `all`,
`argmax`, `sum`) into the composition of common patterns.

Differential Revision: [D29847122](https://our.internmc.facebook.com/intern/diff/D29847122)

[ghstack-poisoned]
